### PR TITLE
Add aten mkldnn conv2d operator

### PR DESCRIPTION
**Pull request by bddppq**
Stack:
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #19633 Add is_mkldnn to at::Tensor&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D15053320/)
&nbsp;&nbsp;&nbsp;&nbsp;:black_circle:&nbsp; **#19204 Add aten mkldnn conv2d operator**&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D14857513/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #19205 Add aten mkldnn ops: relu, max_pool2d and avg_pool2d&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D14850598/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #19206 Add aten mkldnn batch_norm operator&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D14887205/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #19207 Add aten mkldnn add operator&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D14889477/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #19209 Add aten mkldnn view operator&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D14894545/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #19210 Add aten mkldnn linear operator&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D14901641/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #19648 Adjust resnext run script&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D14889481/)

Pull Request resolved: https://github.com/pytorch/pytorch/pull/19204

Differential Revision: [D14857513](https://our.internmc.facebook.com/intern/diff/D14857513/)

### CODE REVIEW

File: aten/src/ATen/native/mkldnn/MKLDNNConversions.cpp [hunk 1]
```diff
@@ -33,6 +34,35 @@ Tensor dense_to_mkldnn(const Tensor& cpu_tensor) {
   return mkldnn_tensor;
 }
 
+Tensor mkldnn_reorder_conv2d_weight(
```
Comment (gchanan) [inline aten/src/ATen/native/mkldnn/MKLDNNConversions.cpp]: what format is this actually changing things into?  And why does this need to be its own function?  Is it because you are trying to cache it?

File: aten/src/ATen/native/Convolution.cpp [hunk 2]
```diff
@@ -301,6 +301,10 @@ at::Tensor conv1d(
 at::Tensor conv2d(
     const Tensor& input, const Tensor& weight, const Tensor& bias,
     IntArrayRef stride, IntArrayRef padding, IntArrayRef dilation, int64_t groups) {
+  if (input.type_id() == MkldnnCPUTensorId()) {
+    return at::mkldnn_convolution(
```
Comment (gchanan) [inline aten/src/ATen/native/Convolution.cpp]: just re-registering from the old PR that this should be folded into the existing dispatch for this op.

File: aten/src/ATen/native/mkldnn/Conv.cpp [hunk 3]
```diff
@@ -35,43 +35,92 @@ std::tuple<at::Tensor,at::Tensor,at::Tensor> mkldnn_convolution_backward(
...
-    auto kernel = dilation[d - 2] * (weight_size[d] - 1) + 1;
-    output_size[d] = (input_size[d] + (2 * padding[d - 2])
-                        - kernel) / stride[d - 2] + 1;
+  const ideep::tensor& x = itensor_from_mkldnn(input);
+  const ideep::tensor& w = itensor_from_mkldnn(weight);
+
+
+  std::vector<int64_t> kernel_size(input.dim());
+  if (groups > 1 && (weight.dim() == input.dim() + 1)) {
+    kernel_size[0] = weight.size(0) * weight.size(1);
```
Comment (gchanan) [inline aten/src/ATen/native/mkldnn/Conv.cpp]: what's going on here?  Why is the `conv_output_size` used here like the other ones but with different parameter names?  What's going on with the kernel_size here?

File: aten/src/ATen/native/mkldnn/Conv.cpp [hunk 4]
```diff
@@ -35,43 +35,92 @@ std::tuple<at::Tensor,at::Tensor,at::Tensor> mkldnn_convolution_backward(
...
 }
 
 at::Tensor mkldnn_convolution(
     const at::Tensor& input, const at::Tensor& weight, const at::Tensor& bias,
     IntArrayRef padding, IntArrayRef stride, IntArrayRef dilation, int64_t groups)
 {
-  auto output = at::empty(conv_output_size(
-    input.sizes(), weight.sizes(), padding, stride, dilation, groups), input.options());
+  if (input.type_id() == MkldnnCPUTensorId()) {
+    return _ideep_conv2d(input, weight, bias, padding, stride, dilation, groups);
```
Comment (gchanan) [inline aten/src/ATen/native/mkldnn/Conv.cpp]: can you explain the difference between these two (besides the tensor type)?  Is it because we call ideep here and mkldnn below?  Should we just use one or the other?

File: aten/src/ATen/native/native_functions.yaml [hunk 5]
```diff
@@ -2484,6 +2484,11 @@
   dispatch:
     CPU: dense_to_mkldnn
 
+- func: mkldnn_reorder_conv2d_weight(Tensor self, int[2] padding=0, int[2] stride=1, int[2] dilation=1, int groups=1) -> Tensor
+  variants: method
```
Comment (gchanan) [inline aten/src/ATen/native/native_functions.yaml]: do we need this to be a method?  can it just be a function in python_module nn?

**Review by gchanan — CHANGES_REQUESTED** [review]

File: aten/src/ATen/native/mkldnn/MKLDNNConversions.cpp (hunk 1 above)
Comment (bddppq) [reply aten/src/ATen/native/mkldnn/MKLDNNConversions.cpp]: This is converting conv weight tensor to a non-public mkldnn format.
Initially I was converting all the tensors with `to_mkldnn` which converts to public mkldnn format (undef IIRC), but then saw ~15% perf gap between aten and the existing c2 ideep integration. After debugging with @Jianhui-Li we found this format mismatch will trigger the implicit reordering during runtime in ideep.

File: aten/src/ATen/native/mkldnn/Conv.cpp (hunk 3 above)
Comment (bddppq) [reply aten/src/ATen/native/mkldnn/Conv.cpp]: This is related `mkldnn_reorder_conv2d_weight`, which converts the conv weights from 4d to 5d.

File: aten/src/ATen/native/mkldnn/Conv.cpp (hunk 4 above)
Comment (bddppq) [reply aten/src/ATen/native/mkldnn/Conv.cpp]: Yes I agree we should use one notation, will rename it to `_mkldnn_conv2d`

File: aten/src/ATen/native/native_functions.yaml (hunk 5 above)
Comment (bddppq) [reply aten/src/ATen/native/native_functions.yaml]: sure

File: aten/src/ATen/native/mkldnn/MKLDNNConversions.cpp (hunk 1 above)
Comment (bddppq) [reply aten/src/ATen/native/mkldnn/MKLDNNConversions.cpp]: Re. reason that I added it as a separate function is because this is very specific for conv weights  and it requires to pass along all the conv params (stride, padding, groups)

File: aten/src/ATen/native/Convolution.cpp [hunk 6]
```diff
@@ -147,11 +147,12 @@ auto ConvParams::use_miopen(const at::Tensor& input) const -> bool {
 
 auto ConvParams::use_mkldnn(const at::Tensor& input) const -> bool {
 #if AT_MKLDNN_ENABLED()
```
Comment (uyongw) [inline aten/src/ATen/native/Convolution.cpp]: IMHO we can remove the AT_MKLDNN_ENABLED switch now.

File: aten/src/ATen/native/Convolution.cpp [hunk 7]
```diff
@@ -147,11 +147,12 @@ auto ConvParams::use_miopen(const at::Tensor& input) const -> bool {
...
-         input.scalar_type() == kFloat && // only on CPU Float Tensors
-         !is_dilated() && // doesn't support dilation
-         !transposed && // or transposed tensors
-         input.ndimension() == 4; // must be in NCHW format
+  return (input.type_id() == MkldnnCPUTensorId()) || // input is mkldnn Tensor
+    (input.type().backend() == at::Backend::CPU &&
+     input.scalar_type() == kFloat && // only on CPU Float Tensors
+     !is_dilated() && // doesn't support dilation
+     !transposed && // or transposed tensors
+     input.ndimension() == 4); // must be in NCHW format
```
Comment (uyongw) [inline aten/src/ATen/native/Convolution.cpp]: For the comment, the original 'NCHW format' does not stand now.
BTW, conv with dilation and ndim=3/4/5 (aka. 1D, 2D, 3D) all supported by mkldnn. But we can do it with a separate PR.

Comment (jgong5) [reply aten/src/ATen/native/Convolution.cpp]: These checks are for the conv on CPU tensor. Probably we just keep what has been done on the original path and put all future opts in the new MKL-DNN tensor path?

Comment (uyongw) [reply aten/src/ATen/native/Convolution.cpp]: Oh, that is for CPU tensor. Then forget what I said :-)

File: aten/src/ATen/native/Convolution.cpp [hunk 8]
```diff
@@ -364,10 +377,19 @@ at::Tensor _convolution(
     bool transposed_, IntArrayRef output_padding_, int64_t groups_,
     bool benchmark, bool deterministic, bool cudnn_enabled) {
 
-  auto input = input_r.contiguous();
+  const bool input_is_mkldnn = (input_r.type_id() == MkldnnCPUTensorId());
+  auto input = input_r;
+  if (!input_is_mkldnn) {
+    input = input.contiguous();
+  }
   auto weight = weight_r;
```
Comment (mingfeima) [inline aten/src/ATen/native/Convolution.cpp]: shall we provide MKLDNN tensor with `contiguous()`?
let `is_contiguous()` return `true` will do the job? In case we have a MKLDNN tensor, it should always be contiguous.

File: aten/src/ATen/native/Convolution.cpp [hunk 9]
```diff
@@ -443,8 +468,14 @@ at::Tensor _convolution(
...
-                                    params.padding, params.stride, params.dilation, params.groups);
+    if (!input_is_mkldnn) {
+      output = at::mkldnn_convolution(input, weight.contiguous(), bias.defined() ? bias.contiguous() : bias,
+                                      params.padding, params.stride, params.dilation, params.groups);
+    } else {
+      // do not call contiguous on mkldnn tensor
+      output = at::mkldnn_convolution(input, weight, bias,
+                                      params.padding, params.stride, params.dilation, params.groups);
+    }
 #endif
```
Comment (mingfeima) [inline aten/src/ATen/native/Convolution.cpp]: also once we have `contiguous()`, we don't have to do if-else here depending on `input_is_mkldnn`, just one call to `at::mkldnn_convolution` is needed

File: aten/src/ATen/native/Convolution.cpp (hunk 8 above)
Comment (jgong5) [reply aten/src/ATen/native/Convolution.cpp]: The question is how we can define sound semantics of `contiguous` for opaque tensors. It does not make sense to have `is_contiguous` return true when `contiguous` is not well defined.

Comment (mingfeima) [reply aten/src/ATen/native/Convolution.cpp]: you can check `TensorTypeId` of course, i did not see any discomfortable to let `MkldnnCPUTensorId` return `true` for `tensor.is_contiguous()`.

File: aten/src/ATen/native/Convolution.cpp [hunk 10]
```diff
@@ -186,10 +187,22 @@ auto ConvParams::is_depthwise(
 }
 
 static void check_shape_forward(const at::Tensor& input,
-                                      const at::Tensor& weight, const at::Tensor& bias,
-                                      const ConvParams& params) {
+                                const at::Tensor& weight, const at::Tensor& bias,
+                                const ConvParams& params, bool input_is_mkldnn) {
   int64_t k = input.ndimension();
   int64_t weight_dim = weight.ndimension();
+  std::vector<int64_t> weight_sizes(k);
```
Comment (dzhulgakov) [inline aten/src/ATen/native/Convolution.cpp]: nit: maybe keep weight_sizes as ArrayRef and have a separate vector for sizes in mkldnn case and then point ArrayRef to that vector?

File: aten/src/ATen/native/Convolution.cpp (hunk 8 above)
Comment (dzhulgakov) [reply aten/src/ATen/native/Convolution.cpp]: I'm a little worried of making contiguous work on MKLDNN as the semantics are not well-defined. Maybe add a wrapper function `contiguous_or_mkldnn` and call it only in this implementation?

File: aten/src/ATen/native/mkldnn/MKLDNNConversions.cpp [hunk 11]
```diff
@@ -33,6 +34,35 @@ Tensor dense_to_mkldnn(const Tensor& cpu_tensor) {
...
+      ideep::convolution_forward::expected_weights_descriptor(
+          w.get_dims(),
+          w.get_data_type(),
+          {stride_vec.cbegin(), stride_vec.cend()},
+          {padding_vec.cbegin(), padding_vec.cend()},
+          {padding_vec.cbegin(), padding_vec.cend()},
+          {dilation_vec.cbegin(), dilation_vec.cend()},
+          groups,
+          ideep::algorithm::convolution_direct);
+  ideep::tensor&& result(desc);
```
Comment (dzhulgakov) [inline aten/src/ATen/native/mkldnn/MKLDNNConversions.cpp]: wait, why are you declaring rvalue ref?

File: test/test_mkldnn.py [hunk 12]
```diff
@@ -75,18 +77,38 @@ def test_detach(self):
...
+            N = torch.randint(3, 10, (1,)).item()
+            C = torch.randint(1, 3, (1,)).item() * groups
+            M = torch.randint(1, 3, (1,)).item() * groups
+            x = torch.randn(N, C, 224, 224, dtype=torch.float32) * 100
+            for bias in [True, False]:
+                conv2d = torch.nn.Conv2d(in_channels=C,
+                                         out_channels=M,
+                                         kernel_size=3,
+                                         stride=2,
+                                         padding=1,
```
[the remark this answers is not in the document]
Comment (dzhulgakov) [reply test/test_mkldnn.py]: yeah, we should probably start introducing hypothesis for cases like this. @gchanan - thoughts?

File: torch/utils/mkldnn.py [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+            param.data = fn(param.data)
+            if param._grad is not None:
+                param._grad.data = fn(param._grad.data)
+
+    for key, buf in module._buffers.items():
+        if buf is not None:
+            module._buffers[key] = fn(buf)
+
+    if isinstance(module, torch.nn.Conv2d):
+        module.weight.data = torch._C._nn.mkldnn_reorder_conv2d_weight(
```
Comment (dzhulgakov) [inline torch/utils/mkldnn.py]: @mingfeima - so in general - is this reordering for inference only? if we do it in training, we'd need to make sure that all optimizers can handle different layout/dimensions and even auxiliary weights like for Adam work well.

File: torch/utils/mkldnn.py [hunk 14]
```diff
@@ -0,0 +1,34 @@
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+import torch
+
+
+def to_mkldnn(module):
```
Comment (gchanan) [inline torch/utils/mkldnn.py]: can you just call _apply?

File: torch/utils/mkldnn.py (hunk 13 above)
Comment (gchanan) [reply torch/utils/mkldnn.py]: can we just do the reorder on the fly in the function?  I'm unclear if the reorder is expensive -- it looks pretty simple, but not sure how it's actually implemented.

File: test/test_mkldnn.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (gchanan) [reply test/test_mkldnn.py]: I don't have any objections to that, although it would be great to start with the existing tests so it doesn't live in its own little world.

File: aten/src/ATen/native/mkldnn/Utils.h [hunk 15]
```diff
@@ -0,0 +1,18 @@
+#pragma once
+
+#include <c10/util/ArrayRef.h>
+#include <vector>
+
+namespace at { namespace native {
+
+std::vector<int64_t> expand_param_if_needed(
```
Comment (gchanan) [inline aten/src/ATen/native/mkldnn/Utils.h]: are these the same as the other Conv implementation versions?  Can we just combine them?

File: aten/src/ATen/native/Convolution.cpp (hunk 7 above)
Comment (gchanan) [reply aten/src/ATen/native/Convolution.cpp]: what's the story with the other checks?  We can't use CPU tensor / MKLDNN algo if it's dilated / transposed / or the input dimension is wrong, but we can if it's an MKLDNN tensor?

And where's the check if it's not conv2d -- what happens if you try to call one of the other algos with an MKLDNN tensor?

**Review by gchanan — CHANGES_REQUESTED** [review]

File: torch/utils/mkldnn.py (hunk 13 above)
Comment (jgong5) [reply torch/utils/mkldnn.py]: @dzhulgakov Pre-reordering weights is still useful for training even though perf gain might not be that obvious, depending on the workloads. It is possible to leave existing optimizers intact to support the opaque tensor as long as it supports all the ops and factory functions used by the optimizers.

File: torch/utils/mkldnn.py (hunk 14 above)
Comment (bddppq) [reply torch/utils/mkldnn.py]: `_apply` doesn't pass the module itself to the callback, but we need it to check whether it's conv (and if yes we need to do extra conversion).

File: torch/utils/mkldnn.py (hunk 13 above)
Comment (bddppq) [reply torch/utils/mkldnn.py]: reorder on the fly (repeatedly) will cause 20% slower in conv (and 15% slower in the whole resnext).

File: torch/utils/mkldnn.py [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+    for child in module.children():
+        to_mkldnn(child)
+
+    def fn(t):
+        if t.is_floating_point():
+            return t.to_mkldnn()
+
+    for param in module._parameters.values():
```
Comment (dzhulgakov) [inline torch/utils/mkldnn.py]: it assumes that all ops work on mkldnn which might not be the case. Maybe we should have two versions: conservative that matches only some modules and full-on that handles all params

File: torch/utils/mkldnn.py (hunk 14 above)
Comment (dzhulgakov) [reply torch/utils/mkldnn.py]: If you just want to match children modules - there's .apply

File: torch/utils/mkldnn.py (hunk 16 above)
Comment (bddppq) [reply torch/utils/mkldnn.py]: Hmm supporting partial conversion is hard I think, you will need to add `to_mkldnn` and `to_dense` at the boundaries.

File: aten/src/ATen/native/Convolution.cpp (hunk 10 above)
Comment (bddppq) [reply aten/src/ATen/native/Convolution.cpp]: Does ArrayRef extend the lifetime of a vector? (If not then we still can not avoid the creation of a vector?)

File: aten/src/ATen/native/Convolution.cpp (hunk 8 above)
Comment (mingfeima) [reply aten/src/ATen/native/Convolution.cpp]: @dzhulgakov Indeed the semantics indicates that `mkldnn` blocked layout should be non-contiguous while at the same time mkldnn function calls just need to know that whether the buffer is a physical contiguous chunk of memory (which is always yes for `mkldnn` tensor).

In case we don't inherit contiguous() for `mkldnn` tensor. Another simple method will be move all `input.contiguous()`, `weight.contiguous()`, `bias.contiguous()` from `_convolution()` to each underlying device implementations: a.k.a `cudnn_convolution()` , `mkldnn_convolution()` etc.
In this case, at the upper level from `_convolution()` will be unified (look cleaner) and inside `mkldnn_convolution` we are free to do something like:
```C
// at::native::mkldnn_convolution()
input = is_input_mkldnn ? input : input.contiguous()
weight = is_weight_mkldnn ? weight : weight.contiguous()
```

and we can do this kind of check ONLY inside mkldnn function calls, so `mkldnn` tensor don't have to inherit `contiguous()`

File: aten/src/ATen/native/Convolution.cpp [hunk 17]
```diff
@@ -343,55 +368,52 @@ at::Tensor convolution(
...
+  auto input = input_r;
+  if (!input_is_mkldnn) {
+    input = input.contiguous();
+  }
   auto weight = weight_r;
   auto bias = bias_r;
   auto k = weight.ndimension();
+  // mkldnn conv2d weights could have been re-ordered to 5d by
+  // mkldnn_reorder_conv2d_weight
+  if (input_is_mkldnn && (k == input.ndimension() + 1)) {
```
Comment (dzhulgakov) [inline aten/src/ATen/native/Convolution.cpp]: can't we just take input.ndimension in all cases?

File: aten/src/ATen/native/Convolution.cpp [hunk 18]
```diff
@@ -343,55 +368,52 @@ at::Tensor convolution(
...
   if (k == 3) {
     params.view1d_as_2d();
     input = view4d(input);
     weight = view4d(weight);
   }
 
-  auto output = at::empty({0}, input.options());
+  Tensor output;
+  if (!input_is_mkldnn) {
+    output = at::empty({0}, input.options());
```
Comment (dzhulgakov) [inline aten/src/ATen/native/Convolution.cpp]: I think it was an unnecessary assignment in the previous code - every single branch below reassigns the `output`

File: aten/src/ATen/native/mkldnn/Conv.cpp [hunk 19]
```diff
@@ -35,43 +35,84 @@ std::tuple<at::Tensor,at::Tensor,at::Tensor> mkldnn_convolution_backward(
...
+    const Tensor& input, const Tensor& weight, const Tensor& bias,
+    at::IntArrayRef padding, at::IntArrayRef stride, at::IntArrayRef dilation, int64_t groups) {
+  const ideep::tensor& x = itensor_from_mkldnn(input);
+  const ideep::tensor& w = itensor_from_mkldnn(weight);
+
+  std::vector<int64_t> kernel_size(input.dim());
+  // mkldnn conv2d weights could have been re-ordered to 5d by
+  // mkldnn_reorder_conv2d_weight
+  if (weight.dim() == input.dim() + 1) {
+    AT_ASSERTM(
```
Comment (dzhulgakov) [inline aten/src/ATen/native/mkldnn/Conv.cpp]: @mingfeima - just curious - does mkldnn tensor for conv stored somewhere internally the original semantic sizes? I'm just worried that we hard-code logic here that might change later as mkldnn does potentially more optimizations

File: aten/src/ATen/native/mkldnn/Conv.cpp [hunk 20]
```diff
@@ -35,43 +35,84 @@ std::tuple<at::Tensor,at::Tensor,at::Tensor> mkldnn_convolution_backward(
...
+  if (input.type_id() == MkldnnCPUTensorId()) {
+    return _mkldnn_conv2d(input, weight, bias, padding, stride, dilation, groups);
+  }
+
+  auto output = at::empty(
+      conv_output_size(
+          input.sizes(), weight.sizes(), padding, stride, dilation),
+      input.options());
 
   auto cpu_engine = CpuEngine::Instance().get_engine();
```
Comment (dzhulgakov) [inline aten/src/ATen/native/mkldnn/Conv.cpp]: we should change the code below using new stuff too (i.e. to_mkldnn -> conv -> to_dense), but it can be a separate PR. Add a TODO?

**Review by dzhulgakov — COMMENTED** [review]
Looks good to me. @gchanan - feel free to do the final pass

File: aten/src/ATen/native/Convolution.cpp (hunk 17 above)
Comment (bddppq) [reply aten/src/ATen/native/Convolution.cpp]: @dzhulgakov nope because for normal cpu and gpu paths, weight ndims is expected to be the same as input ndims.

File: aten/src/ATen/native/mkldnn/Conv.cpp (hunk 20 above)
Comment (bddppq) [reply aten/src/ATen/native/mkldnn/Conv.cpp]: Yes we should.
In fact if you look at the previous commit history I did that, but then somehow this path becomes slower (comparing to the existing direct mkldnn code) :(, I haven't spent time to investigate it yet, but we should definitely do it (in a separate PR).

File: aten/src/ATen/native/Convolution.cpp (hunk 18 above)
Comment (bddppq) [reply aten/src/ATen/native/Convolution.cpp]: good catch

File: aten/src/ATen/native/mkldnn/Conv.cpp (hunk 19 above)
Comment (mingfeima) [reply aten/src/ATen/native/mkldnn/Conv.cpp]: Nope. mkldnn tensor stores only `mkldnn::memory::dims`.

There might be some risks here when generating `output` with `new_with_itensor_mkldnn` in case `mkldnn::memory::dims` is not align with `at::sizes()`. Since the output aten tensor would have sizes equals to mkldnn dims.

So far for CNN modules, i didn't see any problem, since output are all 4D tensors.
For RNN module, it's going to be an issue, since mkldnn dims and aten sizes are not aligned. (mkldnn rnn weight has 5d dims but aten weight has 2d size). Anyway, perhaps i can wrap this up inside ideep so that you don't have to worry about this.

File: aten/src/ATen/native/Convolution.cpp (hunk 10 above)
Comment (gchanan) [reply aten/src/ATen/native/Convolution.cpp]: ArrayRef doesn't extend the lifetime of the vector.

File: aten/src/ATen/native/Convolution.cpp (hunk 18 above)
Comment (gchanan) [reply aten/src/ATen/native/Convolution.cpp]: is this going to be changed?

File: aten/src/ATen/native/Convolution.cpp [hunk 21]
```diff
@@ -203,35 +217,46 @@ static void check_shape_forward(const at::Tensor& input,
...
-  AT_CHECK(weight.size(0) % groups == 0,
+           " at dimension 0, but got weight of size ", weight_sizes, " instead");
+  AT_CHECK(weight_sizes[0] % groups == 0,
            "Given groups=", groups, ", expected weight to be divisible by ",
-           groups, " at dimension 0, but got weight of size ", weight.sizes(),
+           groups, " at dimension 0, but got weight of size ", weight_sizes,
            " instead");
 
+  AT_CHECK(
+      !(input_is_mkldnn && k != 4),
```
Comment (gchanan) [inline aten/src/ATen/native/Convolution.cpp]: is it feasible to move the "non-standard" checks into the mkldnn part?  Obviously you can't do that with contiguous (we can figure out a better solution for that once we have proper dimension order support), but I don't see why things like this need to be here?

File: torch/utils/mkldnn.py (hunk 13 above)
Comment (gchanan) [reply torch/utils/mkldnn.py]: is that slower than non-mkldnn or slower than mkldnn with optimization?

Comment (gchanan) [reply torch/utils/mkldnn.py]: What I'm trying to get at: if the "on the fly" packing is still faster than the dense path, then it makes sense to expose this op independent of pre-packing.  But if it's not, we need to tie this into our greater weight pre-packing story.

File: aten/src/ATen/native/mkldnn/MKLDNNConversions.cpp [hunk 22]
```diff
@@ -33,6 +34,41 @@ Tensor dense_to_mkldnn(const Tensor& cpu_tensor) {
   return mkldnn_tensor;
 }
 
+// Mkldnn tensor has special non-public format for conv2d weights
+// (dense_to_mkldnn only converts dense tensor to mkldnn tensor with
+// public format). Ideep conv kernel will do implicit reorder if the
+// weight is not already in this optimized format. By the time I'm
```
Comment (dzhulgakov) [inline aten/src/ATen/native/mkldnn/MKLDNNConversions.cpp]: Thanks for writing this comment!

By the time -> At the time of writing this note,

File: torch/utils/mkldnn.py [hunk 23]
```diff
@@ -0,0 +1,32 @@
...
+                param.data = t_fn(param.data)
+                if param._grad is not None:
+                    param._grad.data = t_fn(param._grad.data)
+
+        for key, buf in m._buffers.items():
+            if buf is not None:
+                m._buffers[key] = t_fn(buf)
+
+        if isinstance(m, torch.nn.Conv2d):
+            m.weight.data = torch._C._nn.mkldnn_reorder_conv2d_weight(
```
Comment (dzhulgakov) [inline torch/utils/mkldnn.py]: we will probably need to move this call elsewhere in order to define "prepack for inference" interface in the future. Otherwise it's hard to serialize the mkldnn including the particular layout in the file. But it can be done in a separate PR when we get to more API pieces

**Review by dzhulgakov — APPROVED** [review]
I think it's good to go, yay!

File: torch/utils/mkldnn.py (hunk 23 above)
Comment (jgong5) [reply torch/utils/mkldnn.py]: > we will probably need to move this call elsewhere in order to define "prepack for inference" interface in the future. Otherwise it's hard to serialize the mkldnn including the particular layout in the file.

The pre-packing is also useful for training since it avoids extra reorders as well. In order to serialize the weights properly, can we always `to_dense` when the module is serialized?

Comment (dzhulgakov) [reply torch/utils/mkldnn.py]: For training more work would be needed - for example handling mkldnn tensors in optimizer

What I meant about inference is that we'd need a mechanism to bundle instructions to pack the weights after deserialization to mkldnn format (since as you said they'd be stored in as regular to_dense tensors). We have a prototype mechanism for it (https://github.com/pytorch/pytorch/blob/master/torch/jit/quantized.py#L34) but it needs to be improved. Most likely, we'd need to swap out Conv2d module with special MkldnnConv2d module that handles packing after loading

File: aten/src/ATen/native/Convolution.cpp [hunk 24]
```diff
@@ -147,11 +148,12 @@ auto ConvParams::use_miopen(const at::Tensor& input) const -> bool {
 
 auto ConvParams::use_mkldnn(const at::Tensor& input) const -> bool {
 #if AT_MKLDNN_ENABLED()
-  return input.type().backend() == at::Backend::CPU &&
-         input.scalar_type() == kFloat && // only on CPU Float Tensors
-         !is_dilated() && // doesn't support dilation
-         !transposed && // or transposed tensors
-         input.ndimension() == 4; // must be in NCHW format
+  return (input.is_mkldnn()) || // input is mkldnn Tensor
```
Comment (rrkarim) [inline aten/src/ATen/native/Convolution.cpp]: #23042
